#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [29]:
import pandas as pd

import urllib.request
from PIL import Image

## Data

In [16]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("Dataset/train.csv")


## Data exploration

In [4]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    object 
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    object 
 12  Price_euros       912 non-null    float64
dtypes: float64(2), int64(1), object(10)
memory usage: 92.8+ KB


In [7]:
train.head()

id Company                              Product            TypeName  \
0   852    Dell                        Inspiron 5767            Notebook   
1   414    Asus                         ZenBook Flip  2 in 1 Convertible   
2   283  Lenovo                   Ideapad 320-15IKBN            Notebook   
3   669    Acer                             Aspire 3            Notebook   
4  1287      HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook   

   Inches                           ScreenResolution  \
0    17.3                          Full HD 1920x1080   
1    13.3  IPS Panel Full HD / Touchscreen 1920x1080   
2    15.6                          Full HD 1920x1080   
3    15.6                                   1366x768   
4    15.6                                   1366x768   

                          Cpu  Ram     Memory                    Gpu  \
0  Intel Core i7 7500U 2.7GHz  8GB    1TB HDD     AMD Radeon R7 M445   
1  Intel Core i5 7200U 2.5GHz  8GB  256GB SSD  Intel HD Graphics 620   
2  Intel Core i5 7200U 2.5GHz  6GB  256GB SSD  Intel HD Graphics 620   
3    Intel Core i3 6006U 2GHz  4GB  128GB SSD  Intel HD Graphics 520   
4  Intel Core i7 6500U 2.5GHz  6GB    1TB HDD     AMD Radeon R5 M330   

        OpSys  Weight  Price_euros  
0       Linux  2.83kg        865.0  
1  Windows 10  1.27kg        928.0  
2  Windows 10   2.2kg        579.0  
3       Linux   2.1kg        412.0  
4  Windows 10  2.19kg        764.0

## Data processing

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [12]:
X = train.drop(['Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(912, 12)

In [13]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [15]:
X_train

id   Company                                    Product  \
25    748      Dell                              Latitude 5480   
84   1011        HP                              EliteBook 820   
10   1002      Dell                                Vostro 3568   
342   746   Samsung                                 Notebook 9   
890   575  Mediacom                              SmartBook 141   
..    ...       ...                                        ...   
106   397    Lenovo                              ThinkPad L470   
270   714      Dell                              Latitude 5580   
860    85      Dell                              Inspiron 7577   
435   812   Samsung                                 Notebook 9   
102   628      Asus  Q304UA-BHI5T11 (i5-7200U/6GB/1TB/FHD/W10)   

               TypeName  Inches                 ScreenResolution  \
25             Notebook    14.0                         1366x768   
84            Ultrabook    12.5                Full HD 1920x1080   
10             Notebook    15.6                         1366x768   
342           Ultrabook    13.3                Full HD 1920x1080   
890            Notebook    14.0                Full HD 1920x1080   
..                  ...     ...                              ...   
106            Notebook    14.0      IPS Panel Full HD 1920x1080   
270            Notebook    15.6                Full HD 1920x1080   
860              Gaming    15.6      IPS Panel Full HD 1920x1080   
435           Ultrabook    13.3                Full HD 1920x1080   
102  2 in 1 Convertible    13.3  Full HD / Touchscreen 1920x1080   

                             Cpu   Ram                Memory  \
25    Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   
84    Intel Core i5 7200U 2.5GHz   4GB             500GB HDD   
10    Intel Core i3 7100U 2.4GHz   4GB             128GB SSD   
342   Intel Core i7 7500U 2.7GHz  16GB             256GB SSD   
890  Intel Atom x5-Z8350 1.44GHz   4GB              32GB SSD   
..                           ...   ...                   ...   
106   Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   
270   Intel Core i5 7300U 2.6GHz  16GB             512GB SSD   
860  Intel Core i7 7700HQ 2.8GHz  16GB  128GB SSD +  1TB HDD   
435   Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   
102   Intel Core i5 7200U 2.5GHz   6GB               1TB HDD   

                            Gpu       OpSys  Weight  
25        Intel HD Graphics 620  Windows 10   1.6kg  
84        Intel HD Graphics 620  Windows 10  1.26kg  
10        Intel HD Graphics 620  Windows 10  2.18kg  
342       Intel HD Graphics 620  Windows 10  0.81kg  
890           Intel HD Graphics  Windows 10   1.4kg  
..                          ...         ...     ...  
106       Intel HD Graphics 620  Windows 10   1.9kg  
270       Intel HD Graphics 620  Windows 10  1.93kg  
860  Nvidia GeForce GTX 1050 Ti  Windows 10  2.65kg  
435       Intel HD Graphics 620  Windows 10  0.81kg  
102       Intel HD Graphics 620  Windows 10   1.5kg  

[729 rows x 12 columns]

### 3. Asignar el modelo (vacío) a una variable


### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [18]:
X_pred = pd.read_csv("Dataset/test.csv", index_col = 0)
X_pred.head()

Company              Product   TypeName  Inches  \
id                                                     
1184      HP  Pavilion 15-BC000nv   Notebook    15.6   
815   Lenovo       Thinkpad T470s  Ultrabook    14.0   
1244      HP        EliteBook 840   Notebook    14.0   
1121      HP          Spectre Pro   Notebook    13.3   
443   Lenovo    IdeaPad 320-15IAP   Notebook    15.6   

                 ScreenResolution                                   Cpu  Ram  \
id                                                                             
1184  IPS Panel Full HD 1920x1080           Intel Core i7 6700HQ 2.6GHz  6GB   
815             Full HD 1920x1080            Intel Core i7 7500U 2.7GHz  8GB   
1244            Full HD 1920x1080            Intel Core i5 6200U 2.3GHz  4GB   
1121            Full HD 1920x1080            Intel Core i5 6200U 2.3GHz  8GB   
443             Full HD 1920x1080  Intel Pentium Quad Core N4200 1.1GHz  4GB   

         Memory                      Gpu       OpSys  Weight  
id                                                            
1184    1TB HDD  Nvidia GeForce GTX 960M  Windows 10  2.18kg  
815   256GB SSD    Intel HD Graphics 620  Windows 10  1.32kg  
1244  256GB SSD    Intel HD Graphics 520   Windows 7  1.54kg  
1121  256GB SSD    Intel HD Graphics 520  Windows 10  1.16kg  
443   128GB SSD    Intel HD Graphics 505  Windows 10   2.2kg

In [19]:
X_pred.shape

(391, 11)

**¿Por qué da error?** 

In [19]:
predictions_submit = model.predict(X_pred)
predictions_submit

ValueError: could not convert string to float: 'Ideal'

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [ ]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [20]:
X_pred = pd.read_csv("Dataset/test.csv", index_col=0)
X_pred.head()

Company              Product   TypeName  Inches  \
id                                                     
1184      HP  Pavilion 15-BC000nv   Notebook    15.6   
815   Lenovo       Thinkpad T470s  Ultrabook    14.0   
1244      HP        EliteBook 840   Notebook    14.0   
1121      HP          Spectre Pro   Notebook    13.3   
443   Lenovo    IdeaPad 320-15IAP   Notebook    15.6   

                 ScreenResolution                                   Cpu  Ram  \
id                                                                             
1184  IPS Panel Full HD 1920x1080           Intel Core i7 6700HQ 2.6GHz  6GB   
815             Full HD 1920x1080            Intel Core i7 7500U 2.7GHz  8GB   
1244            Full HD 1920x1080            Intel Core i5 6200U 2.3GHz  4GB   
1121            Full HD 1920x1080            Intel Core i5 6200U 2.3GHz  8GB   
443             Full HD 1920x1080  Intel Pentium Quad Core N4200 1.1GHz  4GB   

         Memory                      Gpu       OpSys  Weight  
id                                                            
1184    1TB HDD  Nvidia GeForce GTX 960M  Windows 10  2.18kg  
815   256GB SSD    Intel HD Graphics 620  Windows 10  1.32kg  
1244  256GB SSD    Intel HD Graphics 520   Windows 7  1.54kg  
1121  256GB SSD    Intel HD Graphics 520  Windows 10  1.16kg  
443   128GB SSD    Intel HD Graphics 505  Windows 10   2.2kg

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [24]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([ -260.20029487,   116.70396517,  9699.26794807, ...,
        8903.59284445, 10336.93013348,  -601.00685048])

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [23]:
sample = pd.read_csv("Dataset/sample_submission.csv")

In [24]:
sample.head()

id  Price_euros
0  1184         2158
1   815          511
2  1244         5290
3  1121          622
4   443         1140

In [25]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [26]:
submission = pd.DataFrame({"id": range(len(predictions_submit)), "Price_euros": predictions_submit})

NameError: name 'predictions_submit' is not defined

In [27]:
submission.head()

id  Price_euros
0  1184         2158
1   815          511
2  1244         5290
3  1121          622
4   443         1140

In [28]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [31]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.id.all() == sample.id.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [34]:
chequeator(submission)

You're ready to submit!
